filters/include/pcl/filters/experimental/cartesian_filter.h

```diff
@@ -0,0 +1,150 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/filters/experimental/transform_filter.h>
+
+#include <limits>
+
+namespace pcl {
+namespace experimental {
+
+/** \brief Hash a point to the index of a 2D grid cell.
+ * \param[in] pt the point to hash
+ * \param[in] inverse_leaf_size the array of 1/leaf_size
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] divb_mul_y division multiplier of the second dimension
+ */
+template <typename PointT>
+inline std::size_t
+hashPoint(const PointT& pt,
+          const Eigen::Array4f& inverse_leaf_size,
+          const Eigen::Vector4i& min_b,
+          const std::size_t divb_mul_y)
+{
+  const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size[0]) - min_b[0];
+  const std::size_t ijk1 = std::floor(pt.y * inverse_leaf_size[1]) - min_b[1];
+  return ijk0 + ijk1 * divb_mul_y;
+}
+
+/** \brief Hash a point to the index of a 3D grid cell.
+ * \param[in] pt the point to hash
+ * \param[in] inverse_leaf_size the array of 1/leaf_size
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] divb_mul_y division multiplier of the second dimension
+ * \param[in] divb_mul_z division multiplier of the third dimension
+ */
+template <typename PointT>
+inline std::size_t
+hashPoint(const PointT& pt,
+          const Eigen::Array4f& inverse_leaf_size,
+          const Eigen::Vector4i& min_b,
+          const std::size_t divb_mul_y,
+          const std::size_t divb_mul_z)
+{
+  const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size[0]) - min_b[0];
+  const std::size_t ijk1 = std::floor(pt.y * inverse_leaf_size[1]) - min_b[1];
+  const std::size_t ijk2 = std::floor(pt.z * inverse_leaf_size[2]) - min_b[2];
+  return ijk0 + ijk1 * divb_mul_y + ijk2 * divb_mul_z;
+}
+
+/** \brief Check if the leaf size is too small, given the range of the point cloud
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] inverse_leaf_size the array of 1/leaf_size
+ * \note Return false if hash range is larger than the range of size_t, as the
+ * unsigned int "wrap around" effect will occur during hashing a point.
+ * inverse_leaf_size[2] = 0 for checking 2D grids
+ */
+std::size_t
+checkHashRange(const Eigen::Vector4f& min_p,
+               const Eigen::Vector4f& max_p,
+               const Eigen::Array4f& inverse_leaf_size)
+{
+  const double max_size = static_cast<double>(std::numeric_limits<std::size_t>::max());
+
+  const double px = static_cast<double>(max_p[0]) - static_cast<double>(min_p[0]);
+  const double py = static_cast<double>(max_p[1]) - static_cast<double>(min_p[1]);
+  const double pz = static_cast<double>(max_p[2]) - static_cast<double>(min_p[2]);
+  const double dx = std::floor(px * inverse_leaf_size[0]) + 1;
```

Comment:
The +1 might be meaningless if px * inverse_leaf_size is large enough, specially since uint64 has enough space to be truncated due to the significant digits of double